packages/shadycss/shadow-parts.md

```diff
@@ -0,0 +1,345 @@
+    </x-c>
+  </x-b>
+</x-a>
+```
+
+### Custom properties
+
+The spec behavior is that when a `::part` rule consumes a CSS Custom Property,
+the value of that property is determined at each site where that rule is
+_consumed_, not where the rule was _defined_.
+
+For browsers that don't support native CSS custom properties, ShadyCSS emulates
+this behavior by switching to per-instance styling for any node that receives a
+`::part` rule that consumes a CSS Custom Property. This is identical to the
+behavior already used for normal custom property consumption in ShadyCSS, where
+each such host is given a class that is unique for its set of property values.
+
+Native:
+
+```html
+<x-a>
+  #shadow-root
+    <style>
+      x-b::part(greeting) {
+        color: var(--custom-color);
+      }
+    </style>
+
+    <x-b>
+      #shadow-root
+        <style>
+          :host {
+            --custom-color: red;
+          }
+        </style>
+        <div part="greeting">Hello</div>
+    </x-b>
+
+    <x-b>
+      #shadow-root
+        <style>
+          :host {
+            --custom-color: blue;
+          }
+        </style>
+        <div part="greeting">Hello</div>
+    </x-b>
+</x-a>
+```
+
+ShadyCSS:
+
+```html
+<style>
+  x-a .x-b-1 [shady-part~="x-a:x-b:greeting"] {
+    color: red;
+  }
+  x-a .x-b-2 [shady-part~="x-a:x-b:greeting"] {
+    color: blue;
+  }
+</style>
+<x-a>
+  <x-b class="x-b-1">
+    <div shady-part="x-a:x-b:greeting">Hello</div>
+  </x-b>
+  <x-b class="x-b-2">
+    <div shady-part="x-a:x-b:greeting">Hello</div>
+  </x-b>
+</x-a>
+```
+
+## Performance
+
+If an application does not contain any CSS rules containing `::part` selectors
+(as determined by all calls to `prepareTemplate`), then there is minimal
+overhead to Shadow Parts support, as none of the work below needs to be
+performed.
+
+Once a `::part` rule has been discovered in at least one element template, then
+each time an instance of any element is rendered (on `styleElement`), as well
+as each time an element is added or moved within a shadow root, then a
+`querySelectorAll` is performed to search for nodes with a `part` attribute.
+
+When at least one part node is found in a shadow root, then the element name of
+the host and of the parent host are retrieved, and along with the part names
+themselves, are used to add a `shady-part` attribute to each part node.
+
+In addition, when at least one part node is found in a shadow root, the host is
+checked for an `exportparts` attribute. If found, then _its_ host is checked
+for an `exportparts` attribute, and so on, walking upwards through the
+hierarchy of shadow roots, to construct a part name forwarding map. This map is
+cached for each shadow root, so each shadow root is only walked once. This map
+is then used to add additional `shady-part` attributes to each applicable part
+node.
+
+If a `part` attribute is added or removed from an element, then the appropriate
```

Comment:
By what mechanism is this discovered?

Reply:
I added a special case to the shady DOM `Element.setAttribute` shim:

https://github.com/webcomponents/polyfills/blob/shady-parts-dev/packages/shadydom/src/patches/Element.js#L102